DRAFT: Higher-level Pebble integration into Operator Framework

benhoyt opened the pull request:
This is a draft of what the higher-level Pebble integration into the Operator Framework will look like. Intended usage is something like this:

```python
container = model.unit.containers['pgweb']
container = model.unit.get_container('pgweb')  # same as above

container.add_layer("""
summary: Simple override layer
services:
    srv1:
        override: merge
        environment:
            - var1: val1
""")
container.add_layer({  # supplying a dict works too
    'summary': 'Another layer',
    'services': {
        'srv1': {
            'override': 'merge',
            'environment': my_environ,
        },
    },
})
flattened_config = container.get_config()  # rich pebble.Config object
print(flattened_config.to_yaml())

container.start('foo', 'bar')
container.stop('foo')
container.stop('bar')
container.autostart()
```

Or you can drop down to manually accessing the `pebble.API` instance if you want:

```python
pebble = model.containers['pgweb'].pebble
pebble.stop_services(['postgres', 'redis'])
do_backup()
pebble.start_services(['postgres', 'redis'])
```

Notes:

* There can be multiple workload containers in a single charm, so need to distinguish them by name.
  - see https://github.com/hpidcock/testing-charms/blob/master/pebbletest/metadata.yaml
  - hence the need for model.get_container(name) instead of just model.container
* Unit ~= Pod in K8s

benhoyt committed:
First pass at higher-level API for integrating Pebble

benhoyt committed:
Update higher-level API based on feedback meeting

niemeyer commented on ops/model.py:
```diff
@@ -117,7 +119,11 @@ def get_unit(self, unit_name: str) -> 'Unit':
+            container_names = tuple(self._meta.containers)
+        else:
+            container_names = ()
+        return self._cache.get(Unit, unit_name, container_names)
```
This looks suspect. The primary key of the unit is indeed its name. It's been a while since I went through that code, but do we really want containers as part of the constructor, rather than passing them into that single unit somehow?

benhoyt replied:
Yeah, good call. I knew it wasn't great when I was adding it, but didn't think of a better way. I've cleaned this up now so the constructors/key is only the name again, and the "meta" is attached to the `_ModelCache` instance, which things already have access to. This is clean, and the cache seems like a reasonable place to hang it off of.

niemeyer commented on ops/model.py:
```diff
@@ -271,6 +277,7 @@ def __init__(self, name, backend, cache):
         self._cache = cache
         self._is_our_unit = self.name == self._backend.unit_name
         self._status = None
```
Related to the prior comment, note how there's a bunch of data initialized here instead of passed into the constructor. Again, I'm not sure about this, so take it with a grain of salt, but it's worth looking into it and discussing it at least.

niemeyer commented on ops/model.py:
```diff
@@ -346,6 +353,17 @@ def set_workload_version(self, version: str) -> None:
 
+    @property
+    def containers(self) -> typing.Dict:
+        """Return a dict of (Kubernetes) containers indexed by name."""
```
s/Kubernetes//

Specifically this logic should be agnostic. This is a list of containers associated with the unit, k8s or not.

niemeyer commented on ops/model.py:
```diff
@@ -985,6 +1006,52 @@ def location(self):
 
 
+class Container:
+    """Represents a named (Kubernetes) container in an application.
```
s/Kubernetes//

Also, I believe this is related to the unit rather than application?

niemeyer commented on ops/model.py:
```diff
@@ -985,6 +1006,52 @@ def location(self):
+        self._pebble.stop_services(service_names)
+
+    def add_layer(self, config: typing.Union[str, typing.Dict, 'pebble.Config']):
+        """Dynamically add a service (Pebble) configuration layer.
```
After sleeping on it over the weekend, I really think we want a different name than "config" here, because that already has a very specific meaning in juju.  It doesn't necessarily have to be "setup", though. We just need something nice that isn't config.

benhoyt replied:
Yep, I've changed this to `Setup` to match Pebble's terminology and avoid confusion with Juju config.

niemeyer commented on ops/model.py:
```diff
@@ -985,6 +1006,52 @@ def location(self):
+        """
+        if isinstance(config, (str, dict)):
+            config = pebble.Config(config)
+        self._pebble.add_layer(config)
```
Shouldn't we go the opposite way: if we have something that is not a string, transform it into a string. That enables people to use configurations that are known to Pebble but not to the framework yet, as otherwise it'd fail to parse. It also avoids the unnecessary decode+encode roundtrip.

benhoyt replied:
Good call -- done.

niemeyer commented on ops/model.py:
```diff
@@ -985,6 +1006,52 @@ def location(self):
+            config = pebble.Config(config)
+        self._pebble.add_layer(config)
+
+    def get_config(self) -> 'pebble.Config':
```
Related to above.

benhoyt replied:
This one's a bit trickier, because I think you do want the higher-level `pebble.Setup` object returned here. However, I've made the low-level `PebbleClient.get_config()` function return the raw string, so if a developer really needs that level of control that can dip down to the `unit.pebble.get_config()` function.

niemeyer commented on ops/pebble.py:
```diff
@@ -308,6 +317,82 @@ def __repr__(self):
                 ).format(self=self)
 
 
+class Config:
```
Reminder about the conversation we had: I believe things are all properly namespaced not accounting for the module to be used at all times to distinguish, so that importing the specific names inside a different file will preserve context.

This is true for all the type names, but I won't repeat the comment.

benhoyt replied:
Per email discussion, I've changed to `API` to `PebbleClient`, but mostly left the other names alone -- I highly doubt they'll be referenced outside of the `pebble` module, so IMO we don't need the `Pebble...` prefix on everything.

niemeyer commented on ops/pebble.py:
```diff
@@ -308,6 +317,82 @@ def __repr__(self):
+    https://github.com/canonical/pebble/blob/master/internal/setup/setup.go
+    """
+
+    def __init__(self, raw: Union[str, Dict]):
```
Should that accept None?

niemeyer commented on ops/pebble.py:
```diff
@@ -308,6 +317,82 @@ def __repr__(self):
+
+    def to_yaml(self) -> str:
+        """Convert this config to its YAML representation."""
+        return yaml.dump(self.to_dict(), Dumper=_DefaultDumper)
```
DefaultDumper probably needs a better name. It's not clear why it is being provided (safety), and it's not actually the default since we need to provide it explicitly everywhere.

benhoyt replied:
Good call, changed to `_safe_dumper`.

niemeyer commented on ops/pebble.py:
```diff
@@ -308,6 +317,82 @@ def __repr__(self):
+            'description': self.description,
+            'services': {name: service.to_dict() for name, service in self.services.items()},
+        }
+        raise NotImplementedError
```
Leftover?

niemeyer commented on ops/pebble.py:
```diff
@@ -308,6 +317,82 @@ def __repr__(self):
+    __str__ = to_yaml
+
+
 class API:
```
This should probably be PebbleClient or similar, matching the terminology on the Go side, and being slightly more clear in intent.

benhoyt replied:
Done

niemeyer commented on ops/pebble.py:
```diff
@@ -308,6 +317,82 @@ def __repr__(self):
+
+
 class API:
     """Pebble API client."""
```
Commenting here because I can't comment on code below as it was already there: the error message below isn't following the lowcase convention, and it seems a bit unusual in that it addresses the user directly with an order ("You must"). It's also improper in that it talks to the user about a variable name in the code.

It would be nicer as, for example:

"cannot communicate with pebble socket (no path provided and $PEBBLE not set)"

benhoyt replied:
Fixed. I've also taken the environment variable handling out of the class -- you explicitly pass in a `socket_path`, and the calling code (e.g., the command line test client in pebble.py) does environment variable handling if it needs it.

niemeyer reviewed (changes requested):
Here is a first pass. Please let me know how you feel about those details.

benhoyt committed:
Review feedback

benhoyt committed:
Start of some unit tests

benhoyt committed:
Fix issue with how we're threading "meta" through

benhoyt committed:
Low-level Pebble API client (#468)

A `PebbleClient` that speaks the HTTP-over-UNIX-socket protocol that the Pebble server uses. Unit tests for the above (I'm working on more full-fledged tests using a fake Pebble client that actually talks HTTP, but that'll come in a separate PR).

Note: we're going to rename `pebble.API` to `pebble.PebbleClient` in a subsequent PR, per discussion with Gustavo.

benhoyt committed:
Merge branch 'master' into pebble-integration

Conflicts:
ops/pebble.py
test/test_pebble.py

benhoyt commented:
Rebased on master (now that low-level Pebble PR is merged) and moved to https://github.com/canonical/operator/pull/472
